### PR TITLE
buildLuaPackage: enable __structuredAttrs rocks

### PR DESCRIPTION
**Pull request by teto**
it makes overriding easier, instead of having to know internals to decide which of `sqlite = prev.luaLib.overrideLuarocks prev.sqlite (drv: {` or `sqlite = prev.sqlite.overrideAttrs (drv: {` just use the latter

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by teto** [comment]
cc @Artturin @ncfavier as structured attrs experts could you check/advise if I am doing anything wrong here ? I feel like structuredAttrs could solve many lua related issues and I would like to get this in before the 23.05 release.

**Comment by teto** [comment]
nixpkgs-review gives 2 errors for packages that are already broken on master (nvimpager and a lua library that doesn't support lua.4) I tink that's good to merge @vcunat @figsoda

**Comment by arcnmx** [comment]
This seems to have broken a lot in `buildLuaPackage`?

1. if a `meta` attr isn't passed, it now fails evaluation with `error: attribute 'meta' missing`
2. `passthru` from attrs [is now ignored/overwritten if passed](https://github.com/NixOS/nixpkgs/commit/30a70671f49dbb61abb66a5755398919f81d2d25#diff-55d16b1622e6d8b5d3b387b6553d5141e2639048f6754c1bc6b9fa2f032736f6L210).
3. basically any attrs assigned in the builder now overwrite any passed via `attrs` inputs (including [buildInputs](https://github.com/NixOS/nixpkgs/commit/30a70671f49dbb61abb66a5755398919f81d2d25#diff-55d16b1622e6d8b5d3b387b6553d5141e2639048f6754c1bc6b9fa2f032736f6L119))

I [started trying to fix/revert some of this](https://github.com/NixOS/nixpkgs/compare/master...arcnmx:buildLuaPackage-meta) but I kept discovering more problems and gave up. I think this needs more attention before merging?

**Comment by teto** [comment]
do you have specific examples in mind ? maybe you hit this because you are running out of tree lua modules ? I ran nixpkgs-review on it and was able to build everything. There should be no need to pass a passthru anymore. The meta might be broken indeed.

**Comment by arcnmx** [comment]
> maybe you hit this because you are running out of tree lua modules

Well yes, this applies to anyone who might call or use these functions, now they rigidly only work specifically for the packages tested with. I'm under the impression that there's precedence (and expectation) that builders won't just ignore and blindly override input attrs passed by the package using them - all the code that was in place to prevent that from happening has been removed by this PR.

> There should be no need to pass a passthru anymore

`passthru` is just one of many attrs that a package may want to pass to `mkDerivation` for countless reasons - as a builder is a wrapper around `mkDerivation`, one expects it to be respected if passed.

**Comment by teto** [comment]
do you have some public reference of outside usage . Since so few people contribute to the lua ecosystem I am under the (apparently wrong) impression that no one uses it. 

> I think this needs more attention before merging?

This  PR changed a lot of things with the goal to make it easier to use: this removes the need to know when to call overrideLuarocks and overrideAttrs (tip: just use the latter). It took a lot of work and time to get there, if only because building those packages need CPU and disk storage. I am sorry if this broke your workflow but the fixes are trivial in comparison, especially now that the packages are cached. Do not hesitate to add tests to avoid this kind of regression. We have the distinction between unstable and stable precisely for this reason. And nixpkgs makes it easy to rollback.

File: pkgs/development/interpreters/lua-5/build-lua-package.nix [hunk 1]
```diff
@@ -205,16 +201,14 @@ let
 
   passthru = {
     inherit lua; # The lua interpreter
-    inherit externalDeps;
-    inherit luarocks_content;
-  } // passthru;
+  };
```
Comment (Artturin) [inline pkgs/development/interpreters/lua-5/build-lua-package.nix:204]: `} // attrs.passthru or {};`

**Comment by Artturin** [comment]
@teto passthru is very important even with structuredAttrs to avoid unnecessary attrs becoming a part of the derivation (causing unnecessary rebuilds)

File: pkgs/development/interpreters/lua-5/build-lua-package.nix (hunk 1 above)
Comment (Artturin) [reply pkgs/development/interpreters/lua-5/build-lua-package.nix:204]: also env is important to preserve too

File: pkgs/development/interpreters/lua-5/build-lua-package.nix [hunk 2]
```diff
@@ -167,7 +167,7 @@ let
     wrapLuaPrograms
   '' + attrs.postFixup or "";
 
-  installPhase = attrs.installPhase or ''
```
Comment (arcnmx) [inline pkgs/development/interpreters/lua-5/build-lua-package.nix:170]: > I am sorry if this broke your workflow but the fixes are trivial in comparison

I understand it's an impactful/breaking change to the builder in general, the only part I really take issue with here is that no care or attention was given to preserving input attrs.

The lines with `or` like this one (and all the others: meta, passthru, buildInputs, installPhase, checkPhase, etc) were removed while they could've just been left as-is - this leaves the function incorrect and less flexible than it was before, for no benefit or change. It's just silly to remove lines unrelated to your changes that serve a purpose :disappointed:

**Pull request merged**